### PR TITLE
DNS Deep Dive

### PR DESCRIPTION
**Pull request by Jooho**
What is this PR About?

This is about DNS deep dive. Since openshfit 3.6, the way of SkyDNS is changed and this doc tries to explain how it works now.

How should we test or review this PR?

Is there a relevant Trello card or Github issue open for this?

Issue => #217
Trell => https://trello.com/c/watsufdo/76-dns-deep-dive-in-openshift-36

Who would you like to review this?

cc: @redhat-cop/cant-contain-this

### CODE REVIEW

File: playbooks/fundamentals/dns_deep_dive.adoc [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+_Note: No need to reach the Master from the Node to get SVC DNS data_
+
+
+=== Debugging DNS Flow with tcpdump
+
+_Test env:_
+
+- Master Node
+.. IP: 10.10.181.97
+.. Hostname: dhcp181-97.gsslab.rdu2.redhat.com
```
Comment (lpsantil) [inline playbooks/fundamentals/dns_deep_dive.adoc]: s/gsslab.rdu2.redhat.com/example.com/g

File: playbooks/fundamentals/dns_deep_dive.adoc [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+Node query will access Dnsmasq and forward it to upstream DNS.
+
+```
+# Command
+$ dig dhcp181-97.gsslab.rdu2.redhat.com
+
+# Result
+1 10.10.181.196   10.10.181.196  DNS  … dhcp181-97.gsslab.rdu2.redhat.com OPT
+2 10.10.181.196   10.10.182.21   DNS ... dhcp181-97.gsslab.rdu2.redhat.com OPT
+3 10.10.182.21    10.10.181.196   DNS ...  dhcp181-97.gsslab.rdu2.redhat.com A 10.10.181.97 NS ns01.xxx.redhat.com NS ns02.xxx.redhat.com A x.x.x.x A x.x.x.x OPT
```
Comment (lpsantil) [inline playbooks/fundamentals/dns_deep_dive.adoc]: s/xxx.redhat.com/example.com/g

File: playbooks/fundamentals/dns_deep_dive.adoc [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+---
+= OpenShift Container Platform DNS Deep Dive
+
+Jooho Lee <jlee@redhat.com>
+:toc: macro
+:toc-title:
+
+toc::[]
+
+== DNS in OpenShift?
```
[the remark this answers is not in the document]
Comment (Jooho) [reply playbooks/fundamentals/dns_deep_dive.adoc]: it is subtitle and "OpenShift Container Platform DNS Deep Dive" is main title.
Won't change it

File: playbooks/fundamentals/dns_deep_dive.adoc [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+
+Regarding OpenShift, we are usually talking about two DNS servers.
```
[the remark this answers is not in the document]
Comment (Jooho) [reply playbooks/fundamentals/dns_deep_dive.adoc]: Accepted

File: playbooks/fundamentals/dns_deep_dive.adoc [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+
+- Corporate DNS (outside OCP) for Master internal/external hostname and router wildcard name.
+- OpenShift DNS for communication between internal services.
+
+In this blog, I would like to talk about the latter DNS, which is based on skyDNS.
```
[the remark this answers is not in the document]
Comment (Jooho) [reply playbooks/fundamentals/dns_deep_dive.adoc]: Accepted

File: playbooks/fundamentals/dns_deep_dive.adoc [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+# Access the pod with Pod IP (Container IP - 10.x.x.x)
+$ oc get pod -o wide
+NAME             READY     STATUS    RESTARTS   AGE              IP                NODE
+hello-openshift   1/1         Running             0            5s        10.130.2.13         XXXX
+
+$ curl 10.130.2.13:8080
+Hello OpenShift!
+
+# Tip - use this command to use the Pod IP from your environment
+$ curl $(oc get pod hello-openshift --template '{{ .status.podIP}}):8080
```
[the remark this answers is not in the document]
Comment (Jooho) [reply playbooks/fundamentals/dns_deep_dive.adoc]: Accepted

File: playbooks/fundamentals/dns_deep_dive.adoc [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+NAME              CLUSTER-IP     EXTERNAL-IP   PORT(S)    AGE
+hello-openshift   172.30.12.46   <none>        8080/TCP   5s
+
+----
+
+== Internal DNS: Why do we need it? 
+
+In order to answer this question, I am going to use a simple example. When it comes to communication between applications in the VM world (e.g. VMware, RHV), each application uses the IP address of the guest, which typically does not change. However, in the Docker world, the Docker container gets a new IP address whenever it spins up.
+
+Kubernetes; which orchestrates Docker, needs a static IP for the container. The Kubernetes Service Object; which provides a Service IP, was created to do this, among other things. The Service Object has a hostname and an IP address. The hostname will not change, but the IP address will change if the Service Object is recreated.  With this Service hostname, stable internal communication between pods (services) becomes possible.
```
[the remark this answers is not in the document]
Comment (Jooho) [reply playbooks/fundamentals/dns_deep_dive.adoc]: Grammatically I don't know what is right but it is changed by another reviewer. So you think ';' have to be changed to `,`, I will change it. please let me know your idea.

File: playbooks/fundamentals/dns_deep_dive.adoc [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+- Dnsmasq routes queries for cluster.local and in-addr.arpa to 127.0.0.1:53
+
+image::images/dns_deep_dive_openshift_3.6_dns_1.png[title=Figure 2. DNS Structure for OpenShift 3.6 (master without node)]
+
+
+image::images/dns_deep_dive_openshift_3.6_dns_1.png[title=Figure 3. DNS Structure for OpenShift 3.6 (master with node)]
+
+
+
+== Deep Dive
```
[the remark this answers is not in the document]
Comment (Jooho) [reply playbooks/fundamentals/dns_deep_dive.adoc]: Accepted

File: playbooks/fundamentals/dns_deep_dive.adoc (hunk 1 above)
Comment (Jooho) [reply playbooks/fundamentals/dns_deep_dive.adoc]: Accepted

File: playbooks/fundamentals/dns_deep_dive.adoc (hunk 2 above)
Comment (Jooho) [reply playbooks/fundamentals/dns_deep_dive.adoc]: Accepted

**Comment by Jooho** [comment]
@lpsantil Thanks for your review, I update doc, please review it agagin

**Comment by etsauer** [comment]
@lpsantil can you review again please?

File: playbooks/fundamentals/dns_deep_dive.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (lpsantil) [reply playbooks/fundamentals/dns_deep_dive.adoc]: Ok.  Works for me.

**Comment by lpsantil** [comment]
LGTM.  Great job Jooho!

**Comment by Jooho** [comment]
Sorry for late update. I am going to update the index.adoc then I will remove the doc.

**Comment by JaredBurck** [comment]
thanks @Jooho and great article btw!

**Pull request closed without merging**